### PR TITLE
ARROW-12940: [R] Expose C interface as R6 methods

### PR DESCRIPTION
**Pull request by jonkeane**
Add methods for exporting/importing the data types that are supported in the C-interface

### CODE REVIEW

**Comment by jonkeane** [comment]
For each of `Array`, `RecordBatch`, `RecordBatchReader` (+ `RecordBatchReaderStream`), `Field`, `Schema`, and `DataType` I have added an `export_to_c()` method (which takes a pointer or pointers and must be called from an instantiation of an object (e.g. `arr <- Array$create(); arr$export_to_c(...)`)) and uses `Export*` to expose the object and an `import_from_c()` method (which takes a pointer or pointers and will result in an instantiation of an object of that type (e.g. `new_array <- Array$import_from_c(...)`)).

This should cover all of the types that are available in the C-interface (though I might have missed one). One thing that I haven't done yet is add methods for `Table`s. The py to R and R to py code does something else there, though [I didn't quite follow how it transforms tables so that they can be transported](https://github.com/apache/arrow/blob/master/r/R/python.R#L90-L104). I'm not sure if we need to add that as a method though.

File: r/R/schema.R [hunk 1]
```diff
@@ -136,6 +137,7 @@ Schema <- R6Class("Schema",
   )
 )
 Schema$create <- function(...) schema_(.fields(list2(...)))
+Schema$import_from_c <- function(ptr) ImportSchema(ptr)
```
Comment (nealrichardson) [inline r/R/schema.R:140]:
```suggestion
Schema$import_from_c <- ImportSchema
```

File: r/R/type.R [hunk 2]
```diff
@@ -49,6 +50,8 @@ DataType <- R6Class("DataType",
   )
 )
 
+DataType$import_from_c <- function(ptr) ImportType(ptr)
```
Comment (nealrichardson) [inline r/R/type.R:53]:
```suggestion
DataType$import_from_c <- ImportType
```

File: r/R/record-batch-reader.R [hunk 3]
```diff
@@ -115,6 +116,9 @@ RecordBatchStreamReader$create <- function(stream) {
   assert_is(stream, "InputStream")
   ipc___RecordBatchStreamReader__Open(stream)
 }
+RecordBatchReader$import_from_c <- RecordBatchStreamReader$import_from_c <- function(stream_ptr) {
+  ImportRecordBatchReader(stream_ptr)
+}
```
Comment (nealrichardson) [inline r/R/record-batch-reader.R:121]:
```suggestion
RecordBatchReader$import_from_c <- RecordBatchStreamReader$import_from_c <- ImportRecordBatchReader
```

File: r/R/field.R [hunk 4]
```diff
@@ -59,6 +60,7 @@ Field$create <- function(name, type, metadata) {
   assert_that(missing(metadata), msg = "metadata= is currently ignored")
   Field__initialize(enc2utf8(name), type, TRUE)
 }
+Field$import_from_c <- function(ptr) ImportField(ptr)
```
Comment (nealrichardson) [inline r/R/field.R:63]:
```suggestion
Field$import_from_c <- ImportField
```

File: r/tests/testthat/test-Array.R [hunk 5]
```diff
@@ -819,3 +819,23 @@ test_that("auto int64 conversion to int can be disabled (ARROW-10093)", {
     expect_true(inherits(as.data.frame(batch)$x, "integer64"))
   })
 })
+
+
+test_that("Array to C-interface", {
```
Comment (nealrichardson) [inline r/tests/testthat/test-Array.R:824]: These tests are good, now we have test coverage for the C bridge without requiring pyarrow.

Very tangentially related: what if we changed py_to_r and r_to_py to use export_to_c and import_from_c? Functionally it's the same but I wonder if it reads better since there's more harmony with the pyarrow methods.

File: r/tests/testthat/test-field.R [hunk 6]
```diff
@@ -36,3 +36,15 @@ test_that("Print method for field", {
     "Field\nzz: dictionary<values=string, indices=int32>"
   )
 })
+
+test_that("Field to C-interface", {
+  field <- field("x", time32("s"))
+
+  # the new way, with a scan -> RecordBatchReader
```
Comment (nealrichardson) [inline r/tests/testthat/test-field.R:43]: invalid comment?

File: r/tests/testthat/test-schema.R [hunk 7]
```diff
@@ -174,3 +174,15 @@ test_that("unify_schemas", {
     schema(b = double(), c = bool(), k = utf8())
   )
 })
+
+test_that("Schema to C-interface", {
+  schema <- schema(b = double(), c = bool())
+
+  # the new way, with a scan -> RecordBatchReader
```
Comment (nealrichardson) [inline r/tests/testthat/test-schema.R:181]: another invalid comment?

File: r/tests/testthat/test-data-type.R [hunk 8]
```diff
@@ -411,3 +411,15 @@ test_that("FixedSizeBinary", {
   expect_error(fixed_size_binary("four"))
   expect_error(fixed_size_binary(c(2, 4)))
 })
+
+test_that("DataType to C-interface", {
+  datatype <- timestamp("ms", timezone = "Asia/Pyongyang")
+
+  # the new way, with a scan -> RecordBatchReader
```
Comment (nealrichardson) [inline r/tests/testthat/test-data-type.R:418]: another?

File: r/tests/testthat/test-RecordBatch.R [hunk 9]
```diff
@@ -519,15 +519,57 @@ test_that("ARROW-11769 - grouping preserved in record batch creation", {
...
+  reader_new$export_to_c(stream_ptr_new)
+  circle_new <- RecordBatchStreamReader$import_from_c(stream_ptr_new)
+  tab_from_c_new <- circle_new$read_table()
+  expect_equal(tab, tab_from_c_new)
+})
+
+test_that("RecordBatch to C-interface", {
+  batch <- RecordBatch$create(example_data)
+
+  # the new way, with an array
```
Comment (nealrichardson) [inline r/tests/testthat/test-RecordBatch.R:563]: how is this "new"?

**Review by nealrichardson — CHANGES_REQUESTED** [review]
Thanks, looks mostly good

File: r/tests/testthat/test-RecordBatch.R (hunk 9 above)
Comment (jonkeane) [reply r/tests/testthat/test-RecordBatch.R:563]: This is cruft from part of my testing process, to confirm I was actually testing what I thought I was actually testing. I should have removed this and the others already

File: r/tests/testthat/test-Array.R (hunk 5 above)
Comment (jonkeane) [reply r/tests/testthat/test-Array.R:824]: Yeah, I thought about spinning up a new subprocess of R to pass them back and forth, but this works and is pretty quick and much less complicated to test the functions — and the pyarrow stuff tests the nested process passing already.

I thought about that, and can totally do it — when I did it before with one of the functions I abandoned that path because it didn't make the code any shorter since these wrappers are pretty thing, but let me try again and see how it reads. The harmony/parallelism is probably worth it even if it's not shorter.

**Review by nealrichardson — APPROVED** [review]
